frontend/catalyst/debug/printing.py

```diff
@@ -15,54 +15,90 @@
 """This module enables runtime printing of program values."""
 
 import builtins
+from functools import partial
 
 import jax
 
+from catalyst.api_extensions.callbacks import debug_callback
 from catalyst.jax_primitives import print_p
 from catalyst.tracing.contexts import EvaluationContext
 
 
 # pylint: disable=redefined-builtin
-def print(x, memref=False):
+def print(fmt, *args, **kwargs):
+    """A print function.
+
+    This print function works with python callbacks, so all arguments will be printed in the same
+    way as the :func:`builtins.print` function prints them.
+
+    One difference is that if the first argument is a string, it may be used similarly to an
+    fstring. Like so:
+
+    .. code-block:: python
+
+        @qjit
+        def cir(a, b, c):
+            debug.print("c={c} b={b} a={a}", a=a, b=b, c=c)
+
+        cir(1, 2, 3)
+    >>> c=3 b=2 a=1
```

Comment:
```suggestion
    This print function allows printing of values at runtime, unlike usage of standard Python ``print`` which will print values at program capture/compile time.
    
    ``debug.print`` is a minimal wrapper around :func:`~.debug.callback` which calls Python's ``builtins.print`` function, and thus will use the same formatting styles as Python's built-in print function.
    
    Args:
        fmt (str): The string to be printed. Note that this may also be a
            format string used to format input arguments (for example 
            ``cost={x}``), similar to those permitted by ``str.format``. See 
            the Python docs on
            `string formatting <https://docs.python.org/3/library/stdtypes.html#str.format>`__
            and `format string syntax <https://docs.python.org/3/library/string.html#formatstrings>`__.
        **args: Arguments to be passed to the format string.
        **kwargs: Keyword arguments to be passed to the format string.
        
    .. seealso:: :func:`~.print_memref`

    **Example**

    .. code-block:: python

        @qjit
        def f(a, b, c):
            debug.print("c={c} b={b} a={a}", a=a, b=b, c=c)

    >>> f(1, 2, 3)
    c=3 b=2 a=1
    
    .. note::
    
        Using Python f-strings as the `fmt` string will not work as expected since they will be treated as Python objects.
        This means that array values embedded in them will have their compile-time representation
        printed, instead of actual data.
```